Sources/SwiftProtobuf/AnyMessageStorage.swift

```diff
@@ -0,0 +1,447 @@
+        let c = s[i]
+        i = s.index(after: i)
+        if c == "/" {
+            typeStart = i
+        }
+    }
+
+    return s[typeStart..<s.endIndex]
+}
+
+internal class AnyMessageStorage {
+  var _typeURL: String = ""
+
+  // Computed to do on demand work.
+  var _value: Data {
+    get {
+      if let value = _valueData {
+        return value
+      }
+
+      if let message = _message {
+        do {
+          return try message.serializedData(partial: true)
+        } catch {
+          return Data()
+        }
+      }
+
+      if let contentJSON = _contentJSON, !_typeURL.isEmpty {
+        let encodedTypeName = typeName(fromURL: _typeURL)
+        if let messageType = Google_Protobuf_Any.lookupMessageType(forMessageName: encodedTypeName) {
+          do {
+            // Hack, make an any to use init(unpackingAny:)
+            var any = Google_Protobuf_Any()
+            any.typeURL = _typeURL
+            any._storage._contentJSON = contentJSON
+            any._storage._valueData = nil
+            let m = try messageType.init(unpackingAny: any)
+            return try m.serializedData(partial: true)
+          } catch {
+            return Data()
+          }
+        }
+      }
+
+      return Data()
+    }
+    set {
+      _valueData = newValue
+      _message = nil
+      _contentJSON = nil
+    }
+  }
+
+  // The possible internal states for _value.
+  //
+  // Note: It might make sense to shift to using an enum for internal
+  // state instead to better enforce this; but that also means we could
+  // never got a model were we might also be able to cache the things
+  // we have.
+  var _valueData: Data? = Data()
+  var _message: Message?
+  var _contentJSON: Data?  // Any json parsed from with the @type removed.
+
+  init() {}
+
+  func copy() -> AnyMessageStorage {
+    let clone = AnyMessageStorage()
+    clone._typeURL = _typeURL
+    clone._valueData = _valueData
+    clone._message = _message
+    clone._contentJSON = _contentJSON
+    return clone
+  }
+
+  func unpackTo<M: Message>(target: inout M) throws {
+    if _typeURL.isEmpty {
+      throw AnyUnpackError.emptyAnyField
+    }
+    let encodedType = typeName(fromURL: _typeURL)
+    if encodedType.isEmpty {
+      throw AnyUnpackError.malformedTypeURL
+    }
+    let messageType = typeName(fromMessage: target)
+    if encodedType != messageType {
+      throw AnyUnpackError.typeMismatch
+    }
+    var protobuf: Data?
+    if let message = _message as? M {
+      target = message
+      return
+    }
+
+    if let message = _message {
+      protobuf = try message.serializedData(partial: true)
+    } else if let value = _valueData {
+      protobuf = value
+    }
+    if let protobuf = protobuf {
+      // Decode protobuf from the stored bytes
+      if protobuf.count > 0 {
+        try protobuf.withUnsafeBytes { (p: UnsafePointer<UInt8>) in
+          try target._protobuf_mergeSerializedBytes(from: p, count: protobuf.count, extensions: nil)
```

Comment:
C++ has that global registry of all descriptors/extensions, so they never have to pass a registry.  ☹️